docs/source/miscellaneous/stubs.rst

```diff
@@ -0,0 +1,635 @@
+   #elif (PHP_VERSION_ID >= 80000)
+       zend_declare_typed_property(class_entry, property_prop_name, &property_prop_default_value, ZEND_ACC_PUBLIC, NULL, (zend_type) ZEND_TYPE_INIT_MASK(MAY_BE_STRING));
+   #endif
+       zend_string_release(property_prop_name);
+
+       return class_entry;
+   }
+
+The prepocessor conditions are necessary because ``enum``s and ``readonly`` properties are PHP 8.1 features and consequently,
+they don't exist in PHP 8.0. Therefore, the registration of ``Number`` is completely omitted, while the ``readonly`` flag is
+not added for ``Foo::$prop`` below PHP 8.1 versions.
+
+**********
+Generating information for the optimizer
+**********
+
+A list of functions is maintained for the optimizer in ``Zend/Optimizer/zend_func_infos.h`` containing extra information
+about the return type and the cardinality of the return value. These pieces of information can enable more accurate
+optimizations (i.e. better type inference). Previously, the file was maintained manually, however since PHP 8.1,
+``gen_stub.php`` takes care of this task by passing the ``--generate-optimizer-info`` option.
+
+A function is added to ``zend_func_infos.h`` if either the ``@return`` or the ``@refcount`` PHPDoc tag supplies
+more information than what is available based on the return type declaration. By default, scalar return types have a
+``refcount`` of ``0``, while non-scalar values are ``N``. If a function can only return newly created non-scalar values,
+its ``refcount`` can be set to ``1``.
+
+An example from the built-in functions:
+
+.. code:: php
+
+   /**
+    * @return array<int, string>
+    * @refcount 1
+    */
+   function get_declared_classes(): array {}
+
+Based on which the following func info entry is provided for the optimizer:
+
+.. code:: c
+
+   F1("get_declared_classes", MAY_BE_ARRAY|MAY_BE_ARRAY_KEY_LONG|MAY_BE_ARRAY_OF_STRING),
+
+Please note that the feature is only available for built-in stubs inside php-src, since currently there
+is no way to provide the function list for the optimizer other than overwriting ``zend_func_infos.h`` directly.
+
+Additionally, functions can be evaluated in compile-time if their arguments are known in compile-time and their
+behavior if free from side-effects as well as it is not affected by the global state. Until PHP 8.2, a list of such
+functions was maintained manually in the optimizer. However, since PHP 8.2, the ``@compile-time-eval`` PHPDoc tag
+can be applied to any functions which conform to the above restrictions in order for them to qualify as evaluable in
+compile-time. The feature internally works by adding the ``ZEND_ACC_COMPILE_TIME_EVAL`` function flag.
+
+As of PHP 8.4, the concept of arity-based frameless functions was introduced. This is another optimization technique,
+which results in faster internal function calls by eliminating unnecessary checks for the number of passed parameters
```

Comment:
Mostly, it is about not creating a stack frame, but passing arguments to the function directly via the `ICALL` operands. Hence the name "frameless".